### DISCUSSION TITLE
Improving the dual boot installation experience

### DISCUSSION THREAD

**Discussion by laurensvalk** [pybricks/support#326] [General]
So far, we've had four (!) different ways to install Pybricks on a SPIKE Prime Hub or Robot Inventor Hub.
- Running lengthy scripts in the official app
- Using a placeholder website
- Using pybricksdev
- Building from source and deploying with DFU

These were mainly proof-of-concepts and development tools. We are now starting to look into making one simple tool that is eventually built into [Pybricks Code](https://www.code.pybricks.com/).

To kick things off, we're making some experimental improvements to `pybricksdev` before updating the web app.

Let's use this discussion to collect some ideas and try some things out.

**Comment by laurensvalk** [comment]
@bertlindeman, do you want to give this a try on the latest master of `pybricksdev`?

```
pybricksdev dual-boot path/to/primehub/firmware.zip
```

The expected output should be something like this:

```
Connecting to /dev/ttyACM0
Uploading data to: _pybricks/firmware.bin
Transfer complete!
Uploading data to: _pybricks/__init__.py
Transfer complete!
Uploading data to: _pybricks/install.py
Transfer complete!
Uploading data to: _pybricks/hash.txt
Transfer complete!
>>> from _pybricks.install import install; install()
Starting installation script.
Checking installation files.
Files looking good!
Getting firmware info.
LEGO Firmware version: b'v1.0.00.0000-abcdefg'
Initializing flash for 971768 bytes.
Copying LEGO firmware.
Copying padding.
Copying Pybricks firmware.
Success! The firmware will be installed when it reboots.
Rebooting soon! Please wait.
```

**Reply by BertLindeman** [reply]
Sure, will do it today.

By the example above, I assume the easiest would be on Linux.

**Reply by laurensvalk** [reply]
The USB connectivity is the same as when you tried it a while ago. We just send different things now.

But yeah, feel free to use the system you prefer. Eventually we will switch to a web-app anyway. This test is more about *what* we send to the hub, not *how*.

**Reply by BertLindeman** [reply]
Tried to set it up on windows (missing validators module and other problems), so gave up and switched to Linux Mint.

First: attempt of course is **without a hub** connected:
```OSError: Could not find hub.```
Oke, that is good.

Then:
```pybricksdev dual-boot /home/bert/Downloads/LEGO/firmware\ Build\ #1022/primehub-firmware.zip```

```
Connecting to /dev/ttyACM0
Currently running dual-boot firmware.
Detected firmware:
{'size': 557056, 'version': 'v1.0.06.0034-b0c335b', 'checksum': 1231192444, 'boot_vector': b'\xc9Z\x08\x08'}
Backing up 557056 bytes of original firmware. Progress:
Backup complete

Preparing external flash.
Writing firmware. Progress:
Done! Verifying firmware.
Download succeeded. Rebooting now...
```
Running this on my win10 laptop via an ssh session to the Linux PC in another room.
Went to see the progress on the hub. And that is the neat 5x5 display getting filled. 
So I did not see the progress on the putty session.

Connected to code.pybricks.com.
Version:
('primehub', '3.0.0b6', '7b3d1b42 on 2021-05-01')

A neat way to write the firmware for dual-boot.

Short-press boots LEGO firmware
Long-press boots pybricks firmware.

Sorry, no critical remark 😉 now.

**Reply by laurensvalk** [reply]
Looks like you're on a very old version though :)

Output looks very different now, as shown above.

EDIT: Looks like I did not push the poetry lock. I did now.

**Reply by BertLindeman** [reply]
Maybe one idea then ;-)

Perhaps check that the input file indeed is a valid zip file **before really doing anything**?

Ran:
``` pybricksdev dual-boot ../../tools/debug_memory_movehub.py```

That results in:
```
Connecting to /dev/ttyACM0
Currently running dual-boot firmware.
Detected firmware:
{'size': 557056, 'version': 'v1.0.06.0034-b0c335b', 'checksum': 1231192444, 'boot_vector': b'\xc9Z\x08\x08'}
Backing up 557056 bytes of original firmware. Progress:
Backup complete

Traceback (most recent call last):
  File "/root/.local/bin/pybricksdev", line 8, in <module>
    sys.exit(main())
  File "/root/.local/pipx/venvs/pybricksdev/lib/python3.8/site-packages/pybricksdev/cli/__init__.py", line 327, in main
    asyncio.run(subparsers.choices[args.tool].tool.run(args))
  File "/usr/lib/python3.8/asyncio/runners.py", line 43, in run
    return loop.run_until_complete(main)
  File "/usr/lib/python3.8/asyncio/base_events.py", line 616, in run_until_complete
    return future.result()
  File "/root/.local/pipx/venvs/pybricksdev/lib/python3.8/site-packages/pybricksdev/cli/__init__.py", line 222, in run
    await installer.install(args.firmware.name)
  File "/root/.local/pipx/venvs/pybricksdev/lib/python3.8/site-packages/pybricksdev/repl_installer.py", line 295, in install
    archive = ZipFile(firmware_archive_path)
  File "/usr/lib/python3.8/zipfile.py", line 1269, in __init__
    self._RealGetContents()
  File "/usr/lib/python3.8/zipfile.py", line 1336, in _RealGetContents
    raise BadZipFile("File is not a zip file")
zipfile.BadZipFile: File is not a zip file
```

The hub after this action, looks like this:
![image](https://user-images.githubusercontent.com/8142081/117019699-ef7f6280-acf5-11eb-98de-5dbdc0e076e2.png)

**Reply by BertLindeman** [reply]
> Looks like you're on a very old version though :)
> 
> Output looks very different now, as shown above.

Will have a look at how I got there.
Did a git pull first . . .

**Reply by laurensvalk** [reply]
> Perhaps check

Nah, zero checks. We'll do that when we write the web app so we can throw this away :smile:

**Reply by BertLindeman** [reply]
How to verify what version I am running?
I did a `git pull` in  `pybricksdev` 
and that now stated `Already up to date.`

Reinstalled on my simple user (did it as root before)
```
cd :~/py/pybricks/git_clone/pybricksdev
python3.8 -m pip install --upgrade pip # ensure pip is up to date first
python3.8 -m pip install pipx
python3.8 -m pipx ensurepath
python3.8 -m pipx install pybricksdev
```

Same result as before...
Confusion here.

**Reply by laurensvalk** [reply]
The changes haven't been put in a full release yet. To use it, clone the latest master and run `poetry install` as you did in the past.

I'll try posting more detailed instructions later.

**Reply by BertLindeman** [reply]
> I'll try posting more detailed instructions later

Please wait until I did a retried.

I love python, but hate all the ways you can interfere with it :wink:

**Reply by BertLindeman** [reply]
Nope, I fail . . . . 
```
~/py/pybricks/newclone/pybricksdev$ poetry install
Creating virtualenv pybricksdev in /home/bert/py/pybricks/newclone/pybricksdev/.venv

  ModuleNotFoundError

  No module named 'virtualenv.seed.via_app_data'

  at <frozen importlib._bootstrap>:973 in _find_and_load_unlocked
```
Help appreciated 😄

**Reply by dlech** [reply]
Did you run `poetry env use python3.8` first?

`python3.8` may need to be replaced with the full absolute path depending on how you installed Python 3.8.

**Reply by BertLindeman** [reply]
There must be something else wrong here 😢 
```poetry env use python3.8```
also returns
```No module named 'virtualenv.seed.via_app_data```

**Reply by dlech** [reply]
Reinstall poetry?

**Reply by laurensvalk** [reply]
@BertLindeman -- don't worry too much about it. I thought it was an easy favor as you've [done the exact same once before](https://github.com/pybricks/pybricksdev/issues/14) :smile:  Let's just wait until the next release then.

**Reply by BertLindeman** [reply]
Thanks David.

After `pipx uninstall pybricksdev`

`poetry install`

did work.. running in a fresh clone of pybricksdev

But ` pybricksdev -v` still shows `pybricksdev v1.0.0a4`

So no change!

And the results of ```pybricksdev dual-boot /home/bert/Downloads/LEGO/firmware\ Build\ #1022/primehub-firmware.zip``` are old-style.

**Reply by BertLindeman** [reply]
> 
> 
> @BertLindeman -- don't worry too much about it. I thought it was an easy favor as you've [done the exact same once before](https://github.com/pybricks/pybricksdev/issues/14) 😄 Let's just wait until the next release then.

Oke Laurens. No need to invest more of your time in something I must have forgotten.
Sorry.
But I will persue this quietly in the background 😄

**Reply by laurensvalk** [reply]
> fresh clone of pybricksdev

Did you `git clone` or download a zip? If you cloned it, you can run `git log` to see which commit is currently checked out.

**Reply by laurensvalk** [reply]
I've also found that poetry isn't always great at dealing with updates of such local projects. It tries to be efficient and skip downloads if you already have it, but sometimes this fails. Even `update` often just doesn't work. This usually does:

```
rm -rf .venv poetry.lock
poetry install
```

**Reply by BertLindeman** [reply]
Finally got poetry and pybricksdev on Linux Mint updated.
Now I am not sure if I already report this anywhere. Sorry if duplicate.

Did `git pull` and `poetry install` is OKE now.

The upload of the firmware usually succeeds but sometimes nothing happens just after that.
Maybe a not trapped timeout or such?

But I never got any further than the start of the install.
Like this: 
```
23:49:39 bert@mammoet Linuxmint:~/py/pybricks/git_clone/pybricksdev$ poetry run pybricksdev dual-boot /home/bert/py/pybricks/firmware/primehub-firmware_build_1030.zip
Connecting to /dev/ttyACM0
Uploading data to: _pybricks/firmware.bin
Transfer complete!
Uploading data to: _pybricks/__init__.py
Transfer complete!
Uploading data to: _pybricks/install.py
Transfer complete!
Uploading data to: _pybricks/hash.txt
Transfer complete!
>>> a9d5091d394b961e8e3b5fceb201e905b6e84014df0093b8885089591aeb6a7d46f042b5033e08e303ac7c659373b7af20bb159faf07c9d07741adbf5ab3299bfrom _pybricks.install import install; install()
Traceback (most recent call last):
  File "<stdin>", line 1
SyntaxError: invalid syntax
Traceback (most recent call last):
  File "<string>", line 1, in <module>
  File "/home/bert/py/pybricks/git_clone/pybricksdev/pybricksdev/cli/__init__.py", line 327, in main
    asyncio.run(subparsers.choices[args.tool].tool.run(args))
  File "/usr/lib/python3.8/asyncio/runners.py", line 43, in run
    return loop.run_until_complete(main)
  File "/usr/lib/python3.8/asyncio/base_events.py", line 616, in run_until_complete
    return future.result()
  File "/home/bert/py/pybricks/git_clone/pybricksdev/pybricksdev/cli/__init__.py", line 222, in run
    await installer.install(args.firmware.name)
  File "/home/bert/py/pybricks/git_clone/pybricksdev/pybricksdev/repl_installer.py", line 145, in install
    await self.exec_line("from _pybricks.install import install; install()")
  File "/home/bert/py/pybricks/git_clone/pybricksdev/pybricksdev/repl_installer.py", line 73, in exec_line
    raise ValueError("Failed to execute line: {0}.".format(line))
ValueError: Failed to execute line: from _pybricks.install import install; install().
```
I know you probably are already a step further, @laurensvalk so I will not be disappointed if you ignore this until the next step 😉

**Reply by laurensvalk** [reply]
Hmm, that's interesting -- thanks for trying. I'm not sure why it seems to be printing that hash value.

Would you mind checking the contents of the `_pybricks/hash.txt` on the hub? So in the LEGO app (or just REPL), do something like:

```python
with open("_pybricks/hash.txt", "r") as f:
    print(f.read())
```

If file transfer was successful (I'm not convinced, but who knows), this should be all that's needed to run the installer on the LEGO firmware:

```
from _pybricks.install import install
install()
```

Also, do you know which LEGO firmware you are running?

**Reply by laurensvalk** [reply]
Just made some changes that should make the whole thing a bit more reliable. It also works on Windows now (at least on my machine). In case you want to try it, do `git pull` and `poetry install` to get the latest changes.

**Reply by BertLindeman** [reply]
> 
> 
> Just made some changes that should make the whole thing a bit more reliable. It also works on Windows now (at least on my machine). In case you want to try it, do `git pull` and `poetry install` to get the latest changes.

Sure want to test!
Both on (my) win10 and Linux Mint it seems that nothing happens.
Waited on Linux over 30 minutes and later on win10 10 minutes, nothing ends up in the /_pybricks directory.

Is there a simple debug possible to see what **is** happening? 
Something I could do?


Strange: windows reports different things compared to `poetry install` after `git pull` under linux.
```poetry install```

```
Installing dependencies from lock file
No dependencies to install or update
Installing the current project: pybricksdev (1.0.0-alpha.4)
```
Linux showed:
```
Installing dependencies from lock file
Package operations: 0 installs, 1 update, 0 removals
  • Updating pybricks (3.0.0a3 3dad15a -> 3.0.0a5 3dad15a)
Installing the current project: pybricksdev (1.0.0-alpha.4)
```

The traceback on Linux on pulling the USB-plug:
```
Traceback (most recent call last):
  File "<string>", line 1, in <module>
  File "/home/bert/py/pybricks/git_clone/pybricksdev/pybricksdev/cli/__init__.py", line 327, in main
    asyncio.run(subparsers.choices[args.tool].tool.run(args))
  File "/usr/lib/python3.8/asyncio/runners.py", line 43, in run
    return loop.run_until_complete(main)
  File "/usr/lib/python3.8/asyncio/base_events.py", line 616, in run_until_complete
    return future.result()
  File "/home/bert/py/pybricks/git_clone/pybricksdev/pybricksdev/cli/__init__.py", line 222, in run
    installer.install(args.firmware.name)
  File "/home/bert/py/pybricks/git_clone/pybricksdev/pybricksdev/repl_installer.py", line 150, in install
    self.upload_file('_pybricks/firmware.bin', pybricks_blob)
  File "/home/bert/py/pybricks/git_clone/pybricksdev/pybricksdev/repl_installer.py", line 120, in upload_file
    while not self.is_ready(b'filetransfer started;\r\n'):
  File "/home/bert/py/pybricks/git_clone/pybricksdev/pybricksdev/repl_installer.py", line 50, in is_ready
    self.parse_input()
  File "/home/bert/py/pybricks/git_clone/pybricksdev/pybricksdev/repl_installer.py", line 43, in parse_input
    data = self.serial.read(self.serial.in_waiting)
  File "/home/bert/py/pybricks/git_clone/pybricksdev/.venv/lib/python3.8/site-packages/serial/serialposix.py", line 549, in in_waiting
    s = fcntl.ioctl(self.fd, TIOCINQ, TIOCM_zero_str)
OSError: [Errno 5] Input/output error
```

The traceback on CTRL-C on Win10:
```Traceback (most recent call last):
  File "<string>", line 1, in <module>
  File "C:\Users\bert.grijshuid\Downloads\Lego\Pybricks\pybricksdev\newclone\pybricksdev\pybricksdev\cli\__init__.py", line 327, in main
    asyncio.run(subparsers.choices[args.tool].tool.run(args))
  File "C:\Program Files\WindowsApps\PythonSoftwareFoundation.Python.3.8_3.8.2800.0_x64__qbz5n2kfra8p0\lib\asyncio\runners.py", line 44, in run
    return loop.run_until_complete(main)
  File "C:\Program Files\WindowsApps\PythonSoftwareFoundation.Python.3.8_3.8.2800.0_x64__qbz5n2kfra8p0\lib\asyncio\base_events.py", line 616, in run_until_complete
    return future.result()
  File "C:\Users\bert.grijshuid\Downloads\Lego\Pybricks\pybricksdev\newclone\pybricksdev\pybricksdev\cli\__init__.py", line 222, in run
    installer.install(args.firmware.name)
  File "C:\Users\bert.grijshuid\Downloads\Lego\Pybricks\pybricksdev\newclone\pybricksdev\pybricksdev\repl_installer.py", line 150, in install
    self.upload_file('_pybricks/firmware.bin', pybricks_blob)
  File "C:\Users\bert.grijshuid\Downloads\Lego\Pybricks\pybricksdev\newclone\pybricksdev\pybricksdev\repl_installer.py", line 120, in upload_file
    while not self.is_ready(b'filetransfer started;\r\n'):
  File "C:\Users\bert.grijshuid\Downloads\Lego\Pybricks\pybricksdev\newclone\pybricksdev\pybricksdev\repl_installer.py", line 50, in is_ready
    self.parse_input()
  File "C:\Users\bert.grijshuid\Downloads\Lego\Pybricks\pybricksdev\newclone\pybricksdev\pybricksdev\repl_installer.py", line 43, in parse_input
    data = self.serial.read(self.serial.in_waiting)
  File "C:\Users\bert.grijshuid\Downloads\Lego\Pybricks\pybricksdev\newclone\pybricksdev\.venv\lib\site-packages\serial\serialwin32.py", line 259, in in_waiting
    raise SerialException("ClearCommError failed ({!r})".format(ctypes.WinError()))
serial.serialutil.SerialException: ClearCommError failed (PermissionError(13, 'Het apparaat herkent de opdracht niet.', None, 22))
```

I can see that the `_pybricks` directory **is** created.
There was a LEGO firmware update on opening the mindstroms App.

**Comment by BertLindeman** [comment]
Why does LEGO not allow to copy the log Grr.  So pictures. 

In the LEGO app using an USB connection
I do not understand any of the output on this log.
Obviously, the first program running is **not** your python program . . 

Had to press "run" at least two times:

![image](https://user-images.githubusercontent.com/8142081/117425808-72393500-af23-11eb-924c-3a18dca5bda4.png)
The last message complete:
![image](https://user-images.githubusercontent.com/8142081/117426016-b4627680-af23-11eb-8d25-c36328a838d5.png)

LEGO Firmware : 
![image](https://user-images.githubusercontent.com/8142081/117426209-ee337d00-af23-11eb-9a8a-9446a2077bab.png)

Would it make sense to re-install LEGO firmware and try again?
Or is the situation as is better to debug?

**Reply by laurensvalk** [reply]
I'm a bit puzzeled as to what I'm seeing and why. Although there's no need to per se, this can help you get back to a clean state to simplify things:

- In the regular LEGO app or REPL, do this:

    ```python
    import firmware
    firmware.ext_flash_erase()
    print("I'm done!")
    ```
  This will take several minutes to complete. It erases all files and programs on the hub storage.
- Do a firmware update with the official app. If you are already on the latest Robot Inventor firmware/app, just switch to the SPIKE app instead (or back again) so you'll still get an update.

This doesn't necessarily solve things, but it should help us get to the same starting point.


EDIT: Clarified which is the regular firmware.

**Reply by BertLindeman** [reply]
> This doesn't necessarily solve things, but it should help us get to the same starting point.

No test is complete without a definitive starting point.

Missing the firmware module somehow.
Tried the most obvious modules from `help('modules')`  in `Pybricks Code v1.0.0-beta.15`

``` python
help('modules')
from pybricks import version
print(version)

try:
    import firmware
except:
    pass

try:
    from __main__ import firmware
except:
    pass
try:
    from micropython import firmware
except:
    pass

try:
    from uio import firmware
except:
    pass

try:
    from usys import firmware
except:
    pass

try:
    from builtins import firmware
except:
    pass

try:
    from pybricks import firmware
except:
    pass

firmware.ext_flash_erase()
print("I'm done!")
```
Results in:
```
__main__          micropython       uio               usys
builtins          pybricks          urandom
math              uerrno            uselect
('primehub', '3.0.0b6', '7b3d1b42 on 2021-05-01')
Traceback (most recent call last):
  File "main.py", line 39, in <module>
NameError: name 'firmware' isn't defined
```

**Reply by laurensvalk** [reply]
You have to use the lego app/firmware :)

![image](https://user-images.githubusercontent.com/12326241/117433001-45891b80-af2b-11eb-94cc-1f43882c4c8e.png)

**Reply by BertLindeman** [reply]
I started when you had "in the regular app" 😄 

No matter what python program I enter in the Mindstorms app, it still shows a memory error followed by `creating empty space for backup` like above.

So installing LEGO Spike prime firmware (on the spike prime hub)
Version at `HUB OS 3.0.22`
Probably a bit over the top, but let me do the flash_erase.
``` python
import firmware
firmware.ext_flash_erase()
print("I'm done!")
```
Rather soon after the start the app states to update, so do that and the update ends in: `Something went wrong`
Did not see the print 'I'm done'.

HUB OS now at 0.0.0
So use "reset settings" to reflash the firmware.
Same error: `Something went wrong`

Back to the LEGO Mindstorms app. And that updates the hub fine! (sigh. . )
HUB OS 2.1.4.13

Installing pybricks firmware using beta.pybricks.com (nice tool!)
Now back to ```('primehub', '3.0.0b5', '8f22a995 on 2021-04-13')```

Will retry the pybricksdev dual-boot. (Now or later)
Thank you for the assistance, Laurens. (Off-topic: any new book in the pipeline?)

**Reply by BertLindeman** [reply]
Similar situation as before (on Linux):
```
15:19:52 bert@mammoet Linuxmint:~/py/pybricks/git_clone/pybricksdev$
poetry run pybricksdev dual-boot /home/bert/py/pybricks/firmware/primehub-firmware_build_1030.zip
```
Terminal output:
```Connecting to /dev/ttyACM0
Uploading data to: _pybricks/firmware.bin
Transfer complete!
Uploading data to: _pybricks/__init__.py
Transfer complete!
Uploading data to: _pybricks/install.py
Transfer complete!
Uploading data to: _pybricks/hash.txt
Transfer complete!
>>> a9d5091d394b961e8e3b5fceb201e905b6e84014df0093b8885089591aeb6a7d46f042b5033e08e303ac7c659373b7af20bb159faf07c9d07741adbf5ab3299bfrom _pybricks.install import install; install()
Traceback (most recent call last):
  File "<stdin>", line 1
SyntaxError: invalid syntax
Traceback (most recent call last):
  File "<string>", line 1, in <module>
  File "/home/bert/py/pybricks/git_clone/pybricksdev/pybricksdev/cli/__init__.py", line 327, in main
    asyncio.run(subparsers.choices[args.tool].tool.run(args))
  File "/usr/lib/python3.8/asyncio/runners.py", line 43, in run
    return loop.run_until_complete(main)
  File "/usr/lib/python3.8/asyncio/base_events.py", line 616, in run_until_complete
    return future.result()
  File "/home/bert/py/pybricks/git_clone/pybricksdev/pybricksdev/cli/__init__.py", line 222, in run
    await installer.install(args.firmware.name)
  File "/home/bert/py/pybricks/git_clone/pybricksdev/pybricksdev/repl_installer.py", line 145, in install
    await self.exec_line("from _pybricks.install import install; install()")
  File "/home/bert/py/pybricks/git_clone/pybricksdev/pybricksdev/repl_installer.py", line 73, in exec_line
    raise ValueError("Failed to execute line: {0}.".format(line))
ValueError: Failed to execute line: from _pybricks.install import install; install().
```

In the Mindstorms app:
``` python
import os

# print(os.listdir())
print("Dir of _pybricks\n\t", os.listdir('_pybricks'))

with open("_pybricks/hash.txt", "r") as f:
    print(f.read())
```

Results in:
![image](https://user-images.githubusercontent.com/8142081/117469050-99105f00-af55-11eb-93d1-2cf22f09e7e9.png)

So the `_pybricks` directory is there, but it is **empty**.

**Comment by BertLindeman** [comment]
After power-off of the hub via the center button it auto rebooted and shows this already quite some time:

https://user-images.githubusercontent.com/8142081/117428050-f8567b00-af25-11eb-9ced-e4aead3b439e.mp4

**Reply by laurensvalk** [reply]
That looks like the default program used to remote control the device attached to the hub, this has nothing to do with Pybricks :)

**Comment by laurensvalk** [comment]
@BertLindeman, I went ahead and updated the web based version.

Try this: https://beta.pybricks.com/tmp/

There's no progress on the LEDs. But it's so fast now that maybe we don't need it :rocket:

**Reply by BertLindeman** [reply]
:wink: Uhrm, Using Microsoft Edge to install on Spike Prime hub. After [install] waited 5 minutes without any sign of completion.

**Short extract of the rest here below:**
connect is OKE
Press [install] all that I see happen is the web page reports under `Installation progress`
```Preparing the hub.```

**More details**
Did not expect this to work, but while I was still waiting for completion:
- Pressed ctrl-F5 
- Pressed [connect]
This just works, so assume nothing was going on in the background.
Pressed [install] anyway. No success.
After power-off of the hub re-did. No luck.

Firmware of the hub according to Mindstorms app `2.1.4.13`
Hmm, the app states "update available" but it was already at 2.1.4.13 . . . 
Retry tmp of beta after this Mindstorms update.
The web page reports in all cases:
```
Installation progress
Preparing the hub.
```
No messages on the developer console (not counting ```GET https://beta.pybricks.com/favicon.ico 404```)

So switch to Spike Prime app - update of course ;-)
Firmware now 3.0.22
Similar result.

At this point looked with a Spike app python program to the files in `/_pybricks` 
It is empty.

**Reply by laurensvalk** [reply]
Thanks, this is very helpful! I'm able to reproduce this and will look into a fix.

**Reply by laurensvalk** [reply]
Would you mind going to the old page (beta.pybricks.com), where you can download a backup of your firmware, and share it here?

**Reply by BertLindeman** [reply]
Backup reported as:
```blob:https://beta.pybricks.com/977e59b2-1c67-4c44-9ff7-7eaa8ad4047a```
Was that what you meant?
[EDIT} or you want me to paste the blob here?

**Reply by laurensvalk** [reply]
You should be able to download the `.bin` file half way during the installation process. Can you attach the file here?

**Comment by BertLindeman** [comment]
Had to zip it as a .bin file is of course not supported.
[firmware-backup-v1.1.01.0002-3e5a12.zip](https://github.com/pybricks/support/files/6459614/firmware-backup-v1.1.01.0002-3e5a12.zip)

**Reply by laurensvalk** [reply]
Great, thanks for sharing. This may take a while, so don't hold your breath as I work on a fix :)

**Reply by BertLindeman** [reply]
Easy does it (for me then 😉)

**Reply by laurensvalk** [reply]
Looks like a new version of the MINDSTORMS Robot Inventor app just came out. Do you mind trying it with the firmware that comes with that one?

**Reply by laurensvalk** [reply]
If you're not sure which is the new version --- there should be a new community section in the app with some familiar names.

**Reply by BertLindeman** [reply]
Will have a try. No problem

**Reply by BertLindeman** [reply]
Update in te Mindstorms app sets firmware to ```2.1.4.13" so not very different.
Let me try to find the "new" app

**Reply by BertLindeman** [reply]
"new" app via microsoft store is at 10.1.0
The firmware does not update but is now at 3.1.7.2
Just to be sure stop the app and disconnect the hub and retry.

I do not understand. I did not see an update of the firmware, yet it now is at 3.1.7.2

Well let's try **```beta.pybricks.com/tmp/```**

Good check:
```Failed to connect. Please wait 60 seconds and try again. Make sure that no other apps are using the hub.```
I should have stopped the Mindstorms app.

Nice Community entries Laurens 😄 

And HEY: success.
No need to add extra optical progress things. 
The progress is well shown in the message %

Now installed ```('primehub', '3.0.0b6', '5d46ab6a on 2021-05-11')```

**Reply by BertLindeman** [reply]
Just to be sure I did it correctly:
from ```https://beta.pybricks.com/tmp/```
All well. 
About one minute to upload the firmware and another minute to reboot and install.
Great, no question about it. No uncertainties about: "what is it doing" and "does it do anything"

**Reply by laurensvalk** [reply]
Thanks for testing! We'll probably but this on beta.pybricks.com then! Then the next step will be to get it into Pybricks Code once we officially add SPIKE Prime / Robot Inventor support.

**Reply by BertLindeman** [reply]
[EDIT] changed "my current LEGO firmware" into "the old LEGO firmware"
Confirm the Robot inventor hub on **the old LEGO firmware** does not install the pybricks dual-boot. Same as the Spike hub.
Would we need a check on the lego firmware before trying to install it?

Robot Inventor hub with the current LEGO firmware also installs via beta.pybricks.com/tmp running on Linux Mint 20.1
 

> Thanks for testing! We'll probably but this on beta.pybricks.com then! Then the next step will be to get it into Pybricks Code once we officially add SPIKE Prime / Robot Inventor support.

🎉

**Reply by laurensvalk** [reply]
Yes, I've just added a version note on https://beta.pybricks.com/ (may need to refresh).

In the final version, we'll probably add an automated check to run only on firmware versions that we have tested.

**Reply by laurensvalk** [reply]
> Confirm the Robot inventor hub on my current LEGO firmware does not install the pybricks dual-boot. Same as the Spike hub.

What do you mean with this?

Does it work with the firmware from the latest MINDSTORMS app, or not?

**Reply by BertLindeman** [reply]
> > Confirm the Robot inventor hub on my current LEGO firmware does not install the pybricks dual-boot. Same as the Spike hub.
> 
> What do you mean with this?
> 
> Does it work with the firmware from the latest MINDSTORMS app, or not?

Sorry, the "not" is wrong there.
The latest Mindstorms firmware on the robot Inventor hub is OKE to install the pybricks dual-boot firmware.
The Microsoft App reports the LEGO firmware as ```3.1.7.2```
The Microsoft App states it is ```10.1.0```

**Reply by laurensvalk** [reply]
If it does not not work, then we are not not good! :tada:

**Reply by BertLindeman** [reply]
Updated the [comment](https://github.com/pybricks/support/discussions/326#discussioncomment-725108)

**Reply by laurensvalk** [reply]
About those numbers; they don't mean a whole lot per se. 

What matters is this version:

![image](https://user-images.githubusercontent.com/12326241/117864683-5f14c500-b295-11eb-8193-4fb6ee30bc5a.png)

But it seems that the 1.3.4 / 10.1.0 app forces you to update to that firmware, so giving that version seems good enough for now.

**Reply by BertLindeman** [reply]
> About those numbers; they don't mean a whole lot per se.
> 
> What matters is this version:
> 
> ![image](https://user-images.githubusercontent.com/12326241/117864683-5f14c500-b295-11eb-8193-4fb6ee30bc5a.png)
> 
> But it seems that the 1.3.4 / 10.1.0 app forces you to update to that firmware, so giving that version seems good enough for now.

There should be a simple way for the basic user to verify, or better an automated test.
The version you show here is only visible to the user in the short time that the pybricks firmware is being installed.

**Reply by laurensvalk** [reply]
Thanks. Yes, we’ll make sure this is visible and checked before the users clicks install.

We’ll probably restore our focus on the other hubs for now to get to the first official release, but it’s good to have something in place for robot inventor that at least works in the mean time.
